zcash_primitives/src/transaction/builder.rs

```diff
@@ -393,22 +512,325 @@ impl Progress {
     }
 }
 
+pub struct SaplingBuilder<P: consensus::Parameters> {
+    anchor: Option<bls12_381::Scalar>,
+    value_balance: Amount,
+    spends: Vec<SpendDescriptionInfo>,
+    outputs: Vec<SaplingOutput<P>>,
+    change_address: Option<(OutgoingViewingKey, PaymentAddress)>,
+}
+
+impl<P: consensus::Parameters> SaplingBuilder<P> {
+    fn new() -> Self {
+        SaplingBuilder {
+            anchor: None,
+            value_balance: Amount::zero(),
+            spends: vec![],
+            outputs: vec![],
+            change_address: None,
+        }
+    }
+
+    pub fn value_balance(&self) -> Amount {
+        self.value_balance
+    }
+
+    /// Adds a Sapling note to be spent in this transaction.
+    ///
+    /// Returns an error if the given Merkle path does not have the same anchor as the
+    /// paths for previous Sapling notes.
+    fn add_spend<R: RngCore>(
+        &mut self,
+        mut rng: R,
+        extsk: ExtendedSpendingKey,
+        diversifier: Diversifier,
+        note: Note,
+        merkle_path: MerklePath<Node>,
+    ) -> Result<(), Error> {
+        // Consistency check: all anchors must equal the first one
+        let cmu = Node::new(note.cmu().into());
+        if let Some(anchor) = self.anchor {
+            let path_root: bls12_381::Scalar = merkle_path.root(cmu).into();
+            if path_root != anchor {
+                return Err(Error::AnchorMismatch);
+            }
+        } else {
+            self.anchor = Some(merkle_path.root(cmu).into())
+        }
+
+        let alpha = jubjub::Fr::random(&mut rng);
+
+        self.value_balance += Amount::from_u64(note.value).map_err(|_| Error::InvalidAmount)?;
+
+        self.spends.push(SpendDescriptionInfo {
+            extsk,
+            diversifier,
+            note,
+            alpha,
+            merkle_path,
+        });
+
+        Ok(())
+    }
+
+    /// Adds a Sapling address to send funds to.
+    #[allow(clippy::too_many_arguments)]
+    pub fn add_output<R: RngCore>(
+        &mut self,
+        mut rng: R,
+        params: &P,
+        target_height: BlockHeight,
+        ovk: Option<OutgoingViewingKey>,
+        to: PaymentAddress,
+        value: Amount,
+        memo: Option<MemoBytes>,
+    ) -> Result<(), Error> {
+        let output =
+            SaplingOutput::new_internal(params, target_height, &mut rng, ovk, to, value, memo)?;
+
+        self.value_balance -= value;
+
+        self.outputs.push(output);
+
+        Ok(())
+    }
+
+    /// Sets the Sapling address to which any change will be sent.
+    ///
+    /// By default, change is sent to the Sapling address corresponding to the first note
+    /// being spent (i.e. the first call to [`Builder::add_sapling_spend`]).
+    pub fn send_change_to(&mut self, ovk: OutgoingViewingKey, to: PaymentAddress) {
+        self.change_address = Some((ovk, to));
+    }
```

Comment:
We don't have a concept of change at the level of a sub-builder; we need the full transaction context to figure out the change. This should remain on the top-level builder, and can eventually support specifying addresses from specific pools.